/**:
  ros__parameters:

    map_frame_id: "map"         # global frame (contains fiducials)
    odom_frame_id: "odom"       # odometry frame
    base_frame_id: "base_link"  # robot frame

    scan_topic: "/lance/lidar_scan"
    imu_topic: "/lance/imu"

    tag_usage_mode: 0     # 1: enable, 0: disable, -1: filtered only
    metrics_pub_freq: 10. # frequency for printing stats

    robot_crop_filter:    # points within this bounding box are exluded from processing (in base_link frame)
      min: [ -0.405, -0.6, 0.0 ]   # min corner ^
      max: [ 0.735, 0.6, 0.959 ]   # max corner ^

    trajectory_filter:
      sampling_window_s: 0.6
      min_filter_window_s: 0.4
      thresh:
        avg_linear_error: 0.13
        avg_angular_error: 0.09
        max_linear_deviation: 0.04
        max_angular_deviation: 0.04

    dlo:
      use_timestamps_as_init: false
      gravity_align: false   # enable/disable gravity alignment on startup
      adaptive_params:
        use: false           # whether or not keyframe params scale with "spaciousness"
        lpf_coeff: 0.8
      keyframe:
        thresh_D: 1.    # distance threshold in meters for creating a new keyframe
        thresh_R: 30.   # rotation threshold in degrees for creating a new keyframe
        submap:
          knn: 5   # how many nearest-neightbor poses to extract when building a submap?
          kcv: 5   # convex hull
          kcc: 5   # concave hull
      initial_pose:
        use: true                         # start with the provided pose?
        position: [ 0., 0., 0. ]          # x, y, z
        orientation: [ 1., 0., 0., 0. ]   # w, x, y, z
      voxel_filter:
        scan:
          use: true   # voxelize each input scan?
          res: 0.15   # leaf size
        submap:
          use: false   # voxelize each submap?
          res: 0.15    # leaf size
        adaptive_leaf_size:
          range_coeff: 0.063    # https://www.desmos.com/3d/0eoickmszv
          stddev_coeff: 0.064
          offset: -0.29
          floor: 0.04
          ceil: 0.5
          precision: 0.01
      immediate_filter:
        use: false
        range: 1.0
        thresh_proportion: 0.3
      imu:
        use: true           # integrate imu data to hint GICP?
        use_orientation: true
        calib_time: 3       # calibration time
      gicp:
        num_threads: 4
        min_num_points: 500  # minimum points required
        s2s:
          k_correspondences: 10
          max_correspondence_distance: 0.4
          max_iterations: 32
          transformation_epsilon: 0.01
          euclidean_fitness_epsilon: 0.01
          ransac:
            iterations: 5
            outlier_rejection_thresh: 0.1
        s2m:
          k_correspondences: 15
          max_correspondence_distance: 0.5
          max_iterations: 32
          transformation_epsilon: 0.0009
          euclidean_fitness_epsilon: 0.005
          ransac:
            iterations: 6
            outlier_rejection_thresh: 0.05

    fiducial_detection:
      max_range: 2.
      plane_seg_thickness: 0.01
      ground_seg_thickness: 0.02
      up_vec_max_angular_dev: 0.2
      planes_max_angular_dev: 0.1
      vox_resolution: 0.03
      max_proportion_leftover: 0.05
      min_num_input_points: 100
      min_plane_seg_points: 15

    mapping:
      frustum_search_radius: 0.04   # 0.05/0.009 (realistic/ideal)
      radial_distance_thresh: 0.05  # 0.07/0.01
      delete_delta_coeff: 0.08      # 0.08
      delete_max_range: 3.5         # 3.5
      add_max_range: 3.5            # 3.5
      voxel_size: 0.1               # 0.15

    traversibility:
      chunk_horizontal_range: 3.5
      chunk_vertical_range: 1.
